docs/design/resource-states.md

```diff
@@ -0,0 +1,222 @@
+* It should be clear to users where to look to determine what the current status of their resource is.
+* The status should include information in both normal and failure cases.
+* When a failure occurs, it should be clear to the user that a failure has happened and what the cause of the failure was.
+* When a failure occurs, there should be an error or code that is programatically consumable (for automation, etc).
+* When a failure occurs, there should be an error or text that is human consumable and ideally more informative than the programatically consumable error code.
+* When a failure occurs, it should be clear to the user whether that failure is transient and the operator can self-recover.
+
+### Examining other projects like ASO
+A quick look across the field of projects similar to ASO suggests that there is a relatively standard approach to solving this problem:
+
+* Crossplane reports status through a [Ready condition](https://crossplane.io/docs/v1.3/reference/troubleshoot.html#resource-status-and-conditions)
+* ACK reports status through a [variety of conditions](https://github.com/aws-controllers-k8s/runtime/blob/8191f606c5975d5ba0a1433fab695533980733ba/apis/core/v1alpha1/conditions.go), including `ACK.Adopted`, `ACK.resourceSynced`, `ACK.Terminal`, etc.
+* Cluster API originally used a `phase` and `failureReason`/`failureMessage` pattern, but has since [moved to use conditions](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200506-conditions.md) and is deprecating the old pattern.
+* Pod uses a combination of [phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) and [conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions), including `PodScheduled`, `ContainersReady`, `Initalized`, and `Ready`.
+
+It seems that most of these (and many core Kubernetes resources) use a `Ready` condition to specify the status.
+
+### More on conditions
+These best practices were gathered from [conditions API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) as of July 2021.
+
+> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See #7856 for more details and discussion.
```

Comment:
The `#7876` is resolving as a link to an ASO issue, which doesn't exist - make it an explicit link back into the original repo.

Reply:
Fixed